benchmark/common.js

```diff
@@ -1,205 +1,207 @@
+  options = Object.assign({}, options);
+  for (const key of Object.keys(configArgs)) {
+    options[key] = [ configArgs[key] ];
+  }
+
+  // Perform a breadth-first walk though all options to genereate a
+  // configuration list that contains all combinations.
+  function recursive(keyIndex, prevConfig) {
+    const key = keys[keyIndex];
+    const values = options[key];
+
+    for (const value of values) {
+      if (typeof value !== 'number' && typeof value !== 'string') {
+        throw new TypeError(`configuration "${key}" had type ${typeof value}`);
+      }
+
+      const currConfig = Object.assign({ [key]: value }, prevConfig);
+
+      if (keyIndex + 1 < keys.length) {
+        recursive(keyIndex + 1, currConfig);
+      } else {
+        queue.push(currConfig);
+      }
+    }
+  }
+
+  if (keys.length > 0) {
+    recursive(0, {});
+  } else {
+    queue.push({});
+  }
+
+  return queue;
+};
+
+function hasWrk() {
+  const result = child_process.spawnSync('wrk', ['-h']);
+  if (result.error && result.error.code === 'ENOENT') {
+    console.error('Couldn\'t locate `wrk` which is needed for running ' +
+      'benchmarks. Check benchmark/README.md for further instructions.');
+    process.exit(1);
+  }
 }
 
 // benchmark an http server.
-Benchmark.prototype.http = function(p, args, cb) {
+const WRK_REGEXP = /Requests\/sec:[ \t]+([0-9\.]+)/;
+Benchmark.prototype.http = function(urlPath, args, cb) {
   hasWrk();
-  var self = this;
-  var regexp = /Requests\/sec:[ \t]+([0-9\.]+)/;
-  var url = 'http://127.0.0.1:' + exports.PORT + p;
+  const self = this;
 
-  args = args.concat(url);
+  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
+  args = args.concat(urlFull);
 
-  var out = '';
-  var child = child_process.spawn('wrk', args);
+  const childStart = process.hrtime();
+  const child = child_process.spawn('wrk', args);
+  child.stderr.pipe(process.stderr);
 
+  // Collect stdout
+  let stdout = '';
   child.stdout.setEncoding('utf8');
+  child.stdout.on('data', (chunk) => stdout += chunk);
 
-  child.stdout.on('data', function(chunk) {
-    out += chunk;
-  });
-
-  child.on('close', function(code) {
-    if (cb)
-      cb(code);
+  child.once('close', function(code) {
+    const elapsed = process.hrtime(childStart);
+    if (cb) cb(code);
 
     if (code) {
       console.error('wrk failed with ' + code);
       process.exit(code);
+      return;
     }
-    var match = out.match(regexp);
-    var qps = match && +match[1];
-    if (!qps) {
-      console.error('%j', out);
-      console.error('wrk produced strange output');
+
+    // Extract requests pr second and check for odd results
+    const match = stdout.match(WRK_REGEXP);
+    if (!match || match.length <= 1) {
+      console.error('wrk produced strange output:');
+      console.error(stdout);
       process.exit(1);
+      return;
     }
-    self.report(+qps);
+
+    // Report rate
+    self.report(parseFloat(match[1]), elapsed);
```

Comment:
not that we'd use it, but this'll fail if the string is in hex/binary. instead simply using `+match[1]` will properly coerce those values.